### DISCUSSION TITLE
Any efficient way to pass the frame buffer to the WebView?

### DISCUSSION THREAD

**Discussion by rustdesk** [tauri-apps/tauri#4177] [Q&A] [answered]
https://github.com/rustdesk/rustdesk/discussions/533#discussioncomment-2785244

**Comment by JonasKruckenberg** [comment, marked as answer]
> Yes, I was trying to find out how to use shared memory in Tauri, but sadly I haven't found any related API.

That's because Tauri chose _Message Passing_ as it's IPC architecture not _Shared Memory_. We basically chose to trade performance in exchange for security. So shared memory for zero-copy data exchange is neither supported nor planned. But our current IPC system isn't perfect either, the bad performance with binary data is something we will be working on after v1 and while not zero-copy, I hope we can deliver performance comparable to WebRTC.

**Reply by rustdesk** [reply]
Thanks

**Reply by rustdesk** [reply]
BTW, do you have ETA?

**Reply by JonasKruckenberg** [reply]
> BTW, do you have ETA?

Not at all, currently I'm working on a macOS PoC to see if the approach I have in mind is even feasible.

If you want to stay in the loop on this, check out the `tao-wry` channel on discord. I'll be posting updates/discussing issues there.

**Reply by rustdesk** [reply]
Thanks

**Reply by John0x** [reply]
@JonasKruckenberg are there any updates on this?
I'm interested as well, since something like an image viewer is pretty much impossible with the current approach

**Reply by krishcolligence** [reply]
@JonasKruckenberg with Tauri IPC supporting passing binary messages,, do you know if the performance is good enough to exchange webrtc streams?

**Reply by JPHutchins** [reply]
I am looking for a cross platform GUI that uses HTML CSS and I require efficient binary streams to/from the frontend (AKA would prefer shared memory).  Is Tauri ready for real-time binary data streams?

**Reply by michael-dm** [reply]
interested too, opt-in shared memory would be game-changer for any graphics-heavy app

**Reply by gabrielsalvador** [reply]
also looking for something like this.

**Reply by FabianLars** [reply]
+1 comments are cool and all but if nobody has actual ideas on how to implement shared memory on non-windows platforms those comments won't make the implementation happen faster. :/

That said, we're aware of the importance of IPC performance and we're not exactly happy with it ourselves, even considering the improvements in 2.0, so this topic is always in our minds.

**Reply by JPHutchins** [reply]
I’ll see what I can find.  Is feature parity between all OS critical?  MacOS?

**Reply by FabianLars** [reply]
Nope, it's probably unrealistic, especially with mobile support being part of Tauri now. It'd just be nice to be possible on more than one OS 🙃

**Reply by oganator** [reply]
> +1 comments are cool and all but if nobody has actual ideas on how to implement shared memory on non-windows platforms those comments won't make the implementation happen faster. :/
> 
> That said, we're aware of the importance of IPC performance and we're not exactly happy with it ourselves, even considering the improvements in 2.0, so this topic is always in our minds.


What about using web assembly for shared memory?

**Reply by FabianLars** [reply]
wasm still runs inside the browser sandbox so that won't help us unless i'm missing something?

**Reply by flo-at** [reply]
It's not zero-copy and maybe not even a very smart idea but what about a temporary (maybe memory-mapped) file to exchange the data? Both, the backend and frontend can read files so it should be possible. The question is: is it faster than message passing?

I'm using Flutter right now and there I'm facing the same problem. I have video buffers and complex graphs that I want to render in real-time. Nothing too crazy. 30-60 fps. In Flutter it's possible (but very complex) to implement shared OpenGL/Vulkan/D3D/.. contexts and textures so the backend can render directly into those. But the amount of code needed to do so is crazy. Tauri sounds like a solid option for future projects.

**Reply by FabianLars** [reply]
> Both, the backend and frontend can read files so it should be possible

How can the frontend read that file?

**Reply by flo-at** [reply]
> How can the frontend read that file?

With this API if I understand correctly: https://tauri.app/v1/api/js/fs/

**Reply by FabianLars** [reply]
This api uses the same message passing IPC under the hood. (same for all tauri apis btw)

**Reply by flo-at** [reply]
In that case it doesn't make much sense, of course. Thanks for clarifying!

**Reply by rustdesk** [reply]
We chose Flutter because of this 3 years ago. I thought there will be some change now, but it seems not.

**Reply by flo-at** [reply]
> We chose Flutter because of this 3 years ago. I thought there will be some change now, but it seems not.

We also chose Flutter (mostly) because of this but it comes with its own trade-offs unfortunately.
